### PR TITLE
REF: more explicit dtypes in strings.accessor

### PR DESCRIPTION
**Pull request by jbrockmendel**
Experimenting with #40489 I found that making Series inference behavior more like Index behavior broke a bunch of strings tests.  This makes the strings code robust to that potential change by being more explicit about dtypes.

### CODE REVIEW

File: pandas/core/strings/accessor.py [hunk 1]
```diff
@@ -2145,7 +2157,7 @@ def startswith(self, pat, na=None):
         dtype: bool
         """
         result = self._data.array._str_startswith(pat, na=na)
-        return self._wrap_result(result, returns_string=False)
+        return self._wrap_result(result, returns_string=False, returns_bool=True)
```
Comment (jreback) [inline pandas/core/strings/accessor.py:2160]: can we just change _wrap_result so we don't have 2 keywords? maybe rename to returns_dtype and pass a string?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/strings/accessor.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/strings/accessor.py:2160]: thats tough bc the end-result dtype depends on self._is_string

Comment (jbrockmendel) [reply pandas/core/strings/accessor.py:2160]: Looking at this again it really doesnt look worth the trouble to try to combine these args

File: pandas/core/strings/accessor.py [hunk 2]
```diff
@@ -209,8 +212,12 @@ def _validate(data):
         # see _libs/lib.pyx for list of inferred types
         allowed_types = ["string", "empty", "bytes", "mixed", "mixed-integer"]
 
-        values = getattr(data, "values", data)  # Series / Index
-        values = getattr(values, "categories", values)  # categorical / normal
+        # TODO: avoid kludge for tests.extension.test_numpy
```
Comment (jreback) [inline pandas/core/strings/accessor.py:215]: umm cn you avoid this?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/strings/accessor.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/strings/accessor.py:215]: can revert to use the `getattr` pattern, either way is an anti-pattern

File: pandas/core/strings/accessor.py (hunk 1 above)
Comment (simonjayhawkins) [reply pandas/core/strings/accessor.py:2163]: I don't think either keywords should really be necessary here, the issue is that the arrays passed when expand is True are object arrays of lists/tuples

which method returns bool, where the dtype of the result is not bool?

Comment (jbrockmendel) [reply pandas/core/strings/accessor.py:2163]: > which method returns bool, where the dtype of the result is not bool?

makes sense, ill give this a try

Comment (jbrockmendel) [reply pandas/core/strings/accessor.py:2163]: e.g `isnumeric` we can get object-dtype result of bools and nans

Comment (simonjayhawkins) [reply pandas/core/strings/accessor.py:2163]: the wrapped result would be object in this case? This is for a object dtype array? for StringArray/ArrayString array the return type should always be boolean. object dtype accessors do cast their results, except extract, so if the result is all bools and no nans, the result dtype should be np.bool

Comment (simonjayhawkins) [reply pandas/core/strings/accessor.py:2163]: > the result dtype should be np.bool

the result to be wrapped dtype should be np.bool

Comment (simonjayhawkins) [reply pandas/core/strings/accessor.py:2163]: these conversions are handled in _str_map in ObjectStringArrayMixin

**Review by jreback — APPROVED** [review]

**Pull request merged**